### DISCUSSION TITLE
How is the application-controller suppose to scale and be HA when running as a StatefulSet and managing only the local cluster?

### DISCUSSION THREAD

**Discussion by michaelajr** [argoproj/argo-cd#6310] [General]
I am confused as to the suggested setup of the application controller when argocd is only managing the local cluster. It seems when more replicas are added they just wait to have remote clusters assigned to them. I.e., they do not participate in managing the local cluster. This aligns with what the documentation says the replicas are for. So how would we provide HA for the application-controller managing only the local cluster? What happens when the pod goes down, or the instance the pod is running on goes down, or a rolling update of the cluster needs to cordon and drain the node - and reschedule the pod to another node? Seems there would be some downtime in these cases.

**Comment by michaelajr** [comment]
I can remove the `ARGOCD_CONTROLLER_REPLICAS` env var and all 3 pods do seems to manage the local cluster. But this was buggy as several of my app-of-apps child apps got stuck. They were synced and healthy, and the `argocd app sync` command was stuck waiting on a child app that was fine.

**Reply by jannfis** [reply]
Hey @michaelajr, currently, the `argocd-application-controller` only scales with the number of clusters managed by Argo CD. The scaling can't be used to shard responsibilities of the controller over multiple applications managed on the same cluster.

This is unfortunately a limitation of the current design and makes the application controller the one component that does not provide HA.

**Reply by michaelajr** [reply]
Ah, I see. Ok. Thank you for the response. Keep up the great work!
